client/src/boot/BootRoutes.js

```diff
@@ -109,18 +109,30 @@ class BootRoutes {
   initReactRouter() {
     reactRouteRegister.updateRootRoute({ component: App });
     const rootRoute = reactRouteRegister.getRootRoute();
-    const routes = reactRouteRegister.getChildRoutes().map((route) => (
-      <Route key={route.path} path={route.path} element={<route.component />} />
-    ));
+    const router = createBrowserRouter(
+      createRoutesFromElements(
+          <Route
+            path={rootRoute.path}
+            element={
+              <rootRoute.component>
+                <NavigationBlocker shouldBlockFn={this.shouldConfirmBeforeUnload} blockMessage={this.getUnsavedChangesMessage()} />
+              </rootRoute.component>
+            }
+          >
+            {reactRouteRegister.getChildRoutes().map(
+              (route) => (
+                <Route key={route.path} path={route.path} element={<route.component />} />
+              )
+            )}
+          </Route>
+      ),
+      { basename: joinUrlPaths(Config.get('baseUrl'), Config.get('adminUrl')) }
+    );
 
     ReactDOM.createRoot(document.getElementsByClassName('cms-content')[0]).render(
       <ApolloProvider client={this.client}>
         <ReduxProvider store={this.store}>
-          <BrowserRouter basename={joinUrlPaths(Config.get('baseUrl'), Config.get('adminUrl'))}>
-            <Routes>
-              <Route path={rootRoute.path} element={<rootRoute.component />}>{routes}</Route>
-            </Routes>
-          </BrowserRouter>
+          <RouterProvider router={router} />
         </ReduxProvider>
       </ApolloProvider>
```

Comment:
Ultimately the only differences here are:
- We now have a `RouterProvider` (which is necessary for the new `useBlocker` hook)
- We're passing a component in as a child of the `App` component (rootRoute.component`)